### PR TITLE
Ms 2891

### PR DESCRIPTION
**Pull request by clee-r7**
to test run rspec

### CODE REVIEW

File: lib/metasploit/framework/data_service/proxy/core.rb [hunk 1]
```diff
@@ -158,12 +138,21 @@ def get_data_service
...
-    @usable = false
-    @initialized = false
-    @mutex = Mutex.new()
+  def setup(opts)
+    begin
+      db_manager = opts.delete(:db_manager)
+      if !db_manager.nil?
+        register_data_service(db_manager, true)
+        @usable = true
+      elsif (opts['DatabaseRemoteProcess'])
```
Comment (mkienow-r7) [inline lib/metasploit/framework/data_service/proxy/core.rb]: Style guide - don't use parentheses around the condition of a control expression

File: lib/metasploit/framework/data_service/proxy/core.rb [hunk 2]
```diff
@@ -158,12 +138,21 @@ def get_data_service
   private
   #######
 
-  def initialize
-    @data_services = {}
-    @data_service_id = 0
-    @usable = false
-    @initialized = false
-    @mutex = Mutex.new()
+  def setup(opts)
```
Comment (mkienow-r7) [inline lib/metasploit/framework/data_service/proxy/core.rb:141]: The old initialization code used a mutex to coordinate access to resources. Why was that removed?

File: lib/msf/core/db_manager/host.rb [hunk 3]
```diff
@@ -56,7 +56,7 @@ def each_host(wspace=workspace, &block)
 
   # Exactly like report_host but waits for the database to create a host and returns it.
   def find_or_create_host(opts)
-    host = get_host(opts)
+    host = get_host(opts.clone)
```
Comment (mkienow-r7) [inline lib/msf/core/db_manager/host.rb:59]: Is `opts` cloned here because of the line `wspace = opts.delete(:workspace) || workspace` in the `get_host` method?

**Comment by mkienow-r7** [comment]
If I run the `list_data_services` or `add_data_service` command I get the following error:

```
[-] Error while running command list_data_services: undefined method `instance' for Metasploit::Framework::DataService::DataProxy:Class
Did you mean?  instance_of?

Call stack:
/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:98:in `cmd_list_data_services'
/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:548:in `run_command'
/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:510:in `block in run_single'
/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:504:in `each'
/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:504:in `run_single'
/metasploit-framework/lib/rex/ui/text/shell.rb:206:in `run'
/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
/metasploit-framework/lib/metasploit/framework/command/base.rb:88:in `start'
./msfconsole:48:in `<main>'
```
```
[-] Error while running command add_data_service: undefined method `instance' for Metasploit::Framework::DataService::DataProxy:Class
Did you mean?  instance_of?

Call stack:
/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:114:in `cmd_add_data_service'
/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:548:in `run_command'
/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:510:in `block in run_single'
/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:504:in `each'
/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:504:in `run_single'
/metasploit-framework/lib/rex/ui/text/shell.rb:206:in `run'
/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
/metasploit-framework/lib/metasploit/framework/command/base.rb:88:in `start'
./msfconsole:48:in `<main>'
```

File: lib/metasploit/framework/data_service/proxy/core.rb (hunk 2 above)
Comment (clee-r7) [reply lib/metasploit/framework/data_service/proxy/core.rb:141]: no longer using singleton so there is no initialization

File: lib/msf/core/db_manager/host.rb (hunk 3 above)
Comment (clee-r7) [reply lib/msf/core/db_manager/host.rb:59]: yes

Comment (mkienow-r7) [reply lib/msf/core/db_manager/host.rb:59]: A number of the methods in this module make calls to `opts.delete` are there more location that need to clone the `opts` being passed to those?

Comment (clee-r7) [reply lib/msf/core/db_manager/host.rb:59]: According to the rspec no - I'm not a fan of this delete logic, I don't want add scope to this ticket if you feel the need to investigate further you can add a ticket to do so, I personally don't want to waste any more time on a technology we probably won't carry forward

Comment (mkienow-r7) [reply lib/msf/core/db_manager/host.rb:59]: My concern is inconsistency. We now have a single call that clones `opts`, but given other methods call `opts.delete` are their callers cloning `opts`?

File: lib/metasploit/framework/data_service/proxy/core.rb [hunk 4]
```diff
@@ -15,11 +13,17 @@ module Metasploit
 module Framework
 module DataService
 class DataProxy
-  include Singleton
   include DataProxyAutoLoader
 
   attr_reader :usable
 
+  def initialize(opts = {})
```
Comment (clee-r7) [inline lib/metasploit/framework/data_service/proxy/core.rb:20]: Changes in this file simplify initialization so that initialization doesn't have to be a separate task by removing singleton pattern.  Failing tests were caused by miss matched contexts caused indirectly by the prior initialization logic.

File: lib/msf/core/db_manager/connection.rb [hunk 5]
```diff
@@ -18,7 +18,7 @@ def after_establish_connection
       migrate
 
       # Set the default workspace
-      framework.db.workspace = framework.db.default_workspace
+      self.workspace = self.default_workspace
```
Comment (clee-r7) [inline lib/msf/core/db_manager/connection.rb:21]: removed infinite loop. fw.db -> initialized db manager which calls this method, which called fw.db (loop)

File: lib/msf/core/db_manager/workspace.rb [hunk 6]
```diff
@@ -58,20 +58,23 @@ def workspace_associations_counts()
   end
 
   def delete_all_workspaces()
-    delete_workspaces(workspaces.map(&:name))
+    return delete_workspaces(workspaces.map(&:name))
```
Comment (clee-r7) [inline lib/msf/core/db_manager/workspace.rb:61]: Changes in this file were to fix a test that relied on logging

File: lib/msf/core/framework.rb [hunk 7]
```diff
@@ -199,7 +199,7 @@ def version
   # @return [Metasploit::Framework::DataService::DataProxy]
   def db
     synchronize {
-      @db ||= Metasploit::Framework::DataService::DataProxy.instance
+      @db ||= get_db
```
Comment (clee-r7) [inline lib/msf/core/framework.rb:202]: All this is part of moving initialization to one place while removing the singleton logic and cyclic reference (fw->db, db->fw)

File: lib/msf/core/modules/metadata/store.rb [hunk 8]
```diff
@@ -104,7 +104,7 @@ def configure_user_store
 
   def get_user_store
     store_dir = ::File.join(Msf::Config.config_directory, "store")
-    FileUtils.mkdir(store_dir) if !::File.exist?(store_dir)
+    FileUtils.makedirs(store_dir) if !::File.exist?(store_dir)
```
Comment (clee-r7) [inline lib/msf/core/modules/metadata/store.rb:107]: Fix broken test logic in module cache branch that causes tests to fail when parent branches are wiped

File: lib/msf/ui/console/command_dispatcher/creds.rb [hunk 9]
```diff
@@ -393,7 +393,6 @@ def creds_search(*args)
     }
 
     tbl = Rex::Text::Table.new(tbl_opts)
-    opts = {}
```
Comment (clee-r7) [inline lib/msf/ui/console/command_dispatcher/creds.rb:396]: fixes bug that clear prior opts

File: lib/msf/ui/console/command_dispatcher/db.rb [hunk 10]
```diff
@@ -90,13 +90,11 @@ def active?
           end
 
           def cmd_set_data_service(service_id)
-            data_proxy = Metasploit::Framework::DataService::DataProxy.instance
```
Comment (clee-r7) [inline lib/msf/ui/console/command_dispatcher/db.rb:93]: All changes in this file reflect the cleanup of the singleton logic

File: lib/msf/ui/console/driver.rb [hunk 11]
```diff
@@ -122,7 +122,6 @@ def initialize(prompt = DefaultPrompt, prompt_char = DefaultPromptChar, opts = {
       enstack_dispatcher(dispatcher)
     end
 
-    framework.db.init(framework, opts)
```
Comment (clee-r7) [inline lib/msf/ui/console/driver.rb:125]: clean up unecessary init logic

File: spec/lib/msf/core/auxiliary/cisco_spec.rb [hunk 12]
```diff
@@ -77,7 +77,7 @@ def myworkspace
   
   context '#cisco_ios_config_eater' do
     before(:example) do
-      expect(aux_cisco).to receive(:myworkspace).and_return(workspace)
+      expect(aux_cisco).to receive(:myworkspace).at_least(:once).and_return(workspace)
```
Comment (clee-r7) [inline spec/lib/msf/core/auxiliary/cisco_spec.rb:80]: changes cause workspace to be called more than once

File: spec/support/shared/examples/msf/db_manager/host.rb [hunk 13]
```diff
@@ -5,7 +5,6 @@
   it { is_expected.to respond_to :get_host }
   it { is_expected.to respond_to :has_host? }
   it { is_expected.to respond_to :hosts }
-  it { is_expected.to respond_to :normalize_host }
```
Comment (clee-r7) [inline spec/support/shared/examples/msf/db_manager/host.rb:8]: normalization now done via util logic

File: lib/msf/core/db_manager/workspace.rb (hunk 6 above)
Comment (jbarnett-r7) [reply lib/msf/core/db_manager/workspace.rb:61]: We have a ticket open to fix all of this goliath stuff to use the "official logging methods". Are these the official logging methods that we should be using?
